colossus/src/main/scala/colossus/protocols/http/HttpResponseParser.scala

```diff
@@ -3,68 +3,231 @@ package protocols.http
 
 import core._
 
-import akka.util.ByteString
+import akka.util.{ByteString, ByteStringBuilder}
 
 import colossus.parsing._
 import HttpParse._
 import Combinators._
 import DataSize._
+import controller._
+import service.DecodedResult
+import scala.language.higherKinds
+
+sealed trait ResponseResult 
+object ResponseResult {
+  case class StaticResponse(response: HttpResponse) extends ResponseResult
+  case class StreamResponse(sink: Sink[DataBuffer], response: HttpResponse) extends ResponseResult
+}
 
-
-object HttpResponseParser {
-
+object HttpResponseParser  {
   val DefaultMaxSize: DataSize = 10.MB
 
-  val DefaultQueueSize : Int = 100
-
-  //not totally in love with this extra object allocation, but feel its more straightforward than extractors
-  case class ParsedHeaders(headers : Seq[(String, String)]) extends HttpHeaderUtils
-
-  def static(size: DataSize = DefaultMaxSize) : HttpResponseParser[HttpResponse] = {
-    new HttpResponseParser[HttpResponse](() => maxSize(size, staticResponse))
-  }
+  //TODO: eliminate duplicate code
+  //TODO: Dechunk on static
 
-  def streaming(size : DataSize = DefaultMaxSize, streamBufferSize : Int = DefaultQueueSize) : HttpResponseParser[StreamingHttpResponseHeaderStub] = {
-    new HttpResponseParser[StreamingHttpResponseHeaderStub](() => maxSize(size, streamedResponse(streamBufferSize)))
+  def staticBody(dechunk: Boolean): Parser[DecodedResult.Static[HttpResponse]] = head |> {parsedHead =>
+    parsedHead.transferEncoding match {
+      case TransferEncoding.Identity => parsedHead.contentLength match {
+        case Some(0) | None => const(DecodedResult.Static(HttpResponse(parsedHead, None)))
+        case Some(n) => bytes(n) >> {body => DecodedResult.Static(HttpResponse(parsedHead, Some(body)))}
+      }
+      case _  => chunkedBody >> {body => DecodedResult.Static(HttpResponse(parsedHead, Some(body)))}
+    }
   }
 
-  def staticResponse: Parser[HttpResponse] = firstLine ~ headers |> {case (version, code) ~ headers =>
-    val pHeaders = ParsedHeaders(headers)
-    val clength =  pHeaders.getContentLength
-    val encoding = pHeaders.getEncoding
-    encoding match {
-      case TransferEncoding.Identity => clength match {
-        case Some(0) | None => const(HttpResponse(version, code, headers, ByteString()))
-        case Some(n) => bytes(n) >> {body => HttpResponse(version, code, headers, body)}
+  def streamBody(dechunk: Boolean): Parser[DecodedResult[StreamingHttpResponse]] = head >> {parsedHead =>
+    parsedHead.transferEncoding match {
+      case TransferEncoding.Identity => parsedHead.contentLength match {
+        case Some(0) | None => DecodedResult.Static(StreamingHttpResponse(parsedHead, None))
+        case Some(n) => streamingResponse(parsedHead, Some(n), false)
```

Comment:
If you want to avoid another allocation, you could just a @ Some(n) => streamingResponse(...a..)